libraries/ms-common/src/stm32f0xx/soft_timer.c

```diff
@@ -0,0 +1,173 @@
+        s_soft_timer_elems[i].duration -= curr_time;
+        // Figure out the next timer.
+        if (min_duration >= s_soft_timer_elems[i].duration) {
+          min_time_id = i;
+          min_duration = s_soft_timer_elems[i].duration;
+        }
+      }
+    }
+  }
+  // Set the active timer id.
+  s_active_timer_id = min_time_id;
+}
+
+void soft_timer_init(void) {
+  // Start the PeiphClock
+  RCC_APB1PeriphClockCmd(RCC_APB1Periph_TIM2, ENABLE);
+  // Stop the timer if it was running and ensure only overflows trigger updates.
+  // Also remove the ARR preload.
+  TIM_Cmd(TIM2, DISABLE);
+  TIM_UpdateRequestConfig(TIM2, TIM_UpdateSource_Regular);
+  TIM_ARRPreloadConfig(TIM2, DISABLE);
+
+  // Disable all the timers and forget the last running timer.
+  s_active_timer_id = SOFT_TIMER_MAX_TIMERS;
+  for (uint32_t i = 0; i < SOFT_TIMER_MAX_TIMERS; i++) {
+    s_soft_timer_elems[i].inuse = false;
+  }
+
+  // Enable the interrupts. TIM2 uses IRQ channel 15.
+  stm32f0xx_interrupt_nvic_enable(15, INTERRUPT_PRIORITY_NORMAL);
+  TIM_ITConfig(TIM2, TIM_IT_Update, ENABLE);
+}
+
+StatusCode soft_timer_start(uint32_t duration_us, SoftTimerCallback callback, void *context,
+                            SoftTimerID *timer_id) {
+  // Get the time this was called.
+  uint32_t curr_time = TIM_GetCounter(TIM2);
+
+  // Enable a critical section.
+  bool critical = critical_section_start();
+
+  for (uint32_t i = 0; i < SOFT_TIMER_MAX_TIMERS; i++) {
+    if (!s_soft_timer_elems[i].inuse) {
+      // Look for an empty timer.
+
+      if (s_active_timer_id >= SOFT_TIMER_MAX_TIMERS) {
+        // New timer will be the only timer. Start it.
+        s_active_timer_id = i;
+        prv_start_timer(duration_us);
+      } else if (duration_us < s_soft_timer_elems[s_active_timer_id].duration - curr_time) {
+        // New timer will run be before the active timer. Update and start it. Use as close to now
+        // as possible for the update.
+        prv_soft_timer_update(TIM_GetCounter(TIM2));
+        s_active_timer_id = i;
+        prv_start_timer(duration_us);
+      } else {
+        // Add the current time to cancel what was on the clock when called.
+        duration_us += curr_time;
+      }
+      // Otherwise, the new timer is longest ignore it until later.
+
+      // Actually populate the timer. Do this after the previous steps as this is a critical section
+      // and this timer shouldn't be updated in prv_soft_timer_update.
+      s_soft_timer_elems[i].duration = duration_us;
+      s_soft_timer_elems[i].callback = callback;
+      s_soft_timer_elems[i].context = context;
+      s_soft_timer_elems[i].inuse = true;
+      *timer_id = i;
+      critical_section_end(critical);
+      return STATUS_CODE_OK;
+    }
+  }
+
+  // Out of timers.
+  critical_section_end(critical);
+  return status_msg(STATUS_CODE_RESOURCE_EXHAUSTED, "Out of software timers.");
+}
+
+bool soft_timer_inuse(void) {
```

Comment:
Since you're already keeping track of ``s_active_timer_id`` and ensuring that it is ``SOFT_TIMER_MAX_TIMERS`` only when there are no active timers, it might make sense to just compare those? (I'm a little wary of suggesting that, since while it'll be more efficient compared to iterating through the array, it's significantly less clear to someone debugging through the code. Your call, I guess)